src/pkg/services/m365/api/graph/service_test.go

```diff
@@ -307,3 +320,102 @@ func (suite *GraphIntgSuite) TestAdapterWrap_retriesBadJWTToken() {
 	assert.True(t, IsErrBadJWTToken(err), clues.ToCore(err))
 	assert.Equal(t, 4, retryInc, "number of retries")
 }
+
+// TestAdapterWrap_retriesInvalidRequest tests adapter retries for graph 400
+// invalidRequest errors. It also tests that retries are only done for GET
+// requests.
+func (suite *GraphIntgSuite) TestAdapterWrap_retriesInvalidRequest() {
+	var (
+		t        = suite.T()
+		retryInc = 0
+		// Formulate a graph error response which is parseable to odata error.
+		graphResp = map[string]any{
+			"error": map[string]any{
+				"code":    invalidRequest,
+				"message": "Invalid request",
+				"innerError": map[string]any{
+					"date":              "2024-01-01T18:00:00",
+					"request-id":        "rid",
+					"client-request-id": "cid",
+				},
+			},
+		}
+	)
+
+	serialized, err := json.Marshal(graphResp)
+	require.NoError(t, err, clues.ToCore(err))
+
+	l := int64(len(serialized))
+
+	// Set up a test middleware to always return a graph 400 invalidRequest
+	// response.
+	returnsGraphResp := mwForceResp{
+		alternate: func(req *http.Request) (bool, *http.Response, error) {
+			retryInc++
+
+			header := http.Header{}
+			header.Set("Content-Length", strconv.Itoa(int(l)))
+			header.Set("Content-Type", "application/json")
+
+			resp := &http.Response{
+				Body:          io.NopCloser(bytes.NewReader(serialized)),
+				ContentLength: l,
+				Header:        header,
+				Proto:         req.Proto,
+				Request:       req,
+				StatusCode:    http.StatusBadRequest, // Required retry condition
+			}
+
+			return true, resp, nil
+		},
+	}
+
+	adpt, err := CreateAdapter(
+		suite.credentials.AzureTenantID,
+		suite.credentials.AzureClientID,
+		suite.credentials.AzureClientSecret,
+		count.New(),
+		appendMiddleware(&returnsGraphResp))
+	require.NoError(t, err, clues.ToCore(err))
+
+	table := []struct {
+		name            string
+		apiRequest      func(t *testing.T, ctx context.Context) error
+		expectedRetries int
+	}{
+		{
+			name: "GET request, retried",
+			apiRequest: func(t *testing.T, ctx context.Context) error {
+				_, err = NewService(adpt).Client().Users().Get(ctx, nil)
+				return err
+			},
+			expectedRetries: 4,
+		},
+		{
+			name: "POST request, no retry",
+			apiRequest: func(t *testing.T, ctx context.Context) error {
+				u := models.NewUser()
+
+				cfg := users.UsersRequestBuilderPostRequestConfiguration{}
+
+				_, err = NewService(adpt).Client().Users().Post(ctx, u, &cfg)
+				return err
+			},
+			expectedRetries: 1,
+		},
+	}
+
+	for _, test := range table {
+		suite.Run(test.name, func() {
+			tt := suite.T()
```

Comment:
You don't have to worry about clobbering here, since you're creating a new var.
```suggestion
			t := suite.T()
```

Reply:
We have `t` defined above at L329. Didn't want to shadow that var name.